### PR TITLE
Fix our x86 test execution

### PR DESCRIPTION
**Pull request by jaredpar**
The transition to `dotnet test` seems to have broken our unit tests that
depend on executing on an x86 architecture. This fixes that by using the
`--arch` argument to the `dotnet test` execution.

This was made more general to facilitate our future arm64 work in this
area

### CODE REVIEW

**Comment by jaredpar** [comment]
Verified locally this does cause the x86 test runner to be used in `-testArch x86`

**Comment by jaredpar** [comment]
@dotnet/roslyn-infrastructure PTAL

**Review by 333fred — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Is there anything else we should do to verify here? For example, temporarily insert a test which has different behaviors depending on `sizeof(nint)` and check that we get the expected behavior in each test job?

**Review by JoeRobich — APPROVED** [review]

**Review by RikkiGibson — APPROVED** [review]

**Comment by jaredpar** [comment]
> Is there anything else we should do to verify here? For example, temporarily insert a test which has different behaviors depending on sizeof(nint) and check that we get the expected behavior in each test job?

I could not think of a simple way to verify this on x86 / x64. The tests are only compiled once hence not clear how we can have tests that validate specific behaviors that we can query on. 

The good news is that once we go to arm64 this gets trivial to verify. Our test leg times out even when running minimal tests if we accidentally use x64 emulation

**Comment by RikkiGibson** [comment]
I opened #60324 to test in the way I described. I don't really anticipate having an ongoing test for this, it seems too much like "assert that X is X", but thought a sanity test at the time we introduce it would be useful.

**Comment by jaredpar** [comment]
Great news, I was wrong. Our tests were actually successfully executing on x86. 
Bad news, they were not executing on x64 ... ever

File: src/EditorFeatures/CSharpTest2/EmbeddedLanguages/Json/CSharpJsonTests.cs [hunk 1]
```diff
@@ -12,6 +12,7 @@
 using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
 using Microsoft.CodeAnalysis.Features.EmbeddedLanguages.Json;
 using Newtonsoft.Json.Linq;
+using Roslyn.Test.Utilities;
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/EmbeddedLanguages/Json/CSharpJsonTests.cs:15]:
```suggestion
```

File: src/Tools/Source/RunTests/ProcessTestExecutor.cs [hunk 2]
```diff
@@ -72,6 +72,7 @@ void MaybeAddSeparator(char separator = '|')
                 }
             }
 
+            builder.Append($@" --arch {assemblyInfo.Architecture}");
```
Comment (sharwell) [inline src/Tools/Source/RunTests/ProcessTestExecutor.cs:75]: 😅

**Review by sharwell — APPROVED** [review]

File: src/EditorFeatures/CSharpTest2/EmbeddedLanguages/Json/CSharpJsonTests.cs [hunk 3]
```diff
@@ -384,6 +385,11 @@ public void TestDeepRecursion()
 [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
 [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
 [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
+[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
```
Comment (jaredpar) [inline src/EditorFeatures/CSharpTest2/EmbeddedLanguages/Json/CSharpJsonTests.cs:392]:
```suggestion
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[
```

File: eng/build.ps1 [hunk 4]
```diff
@@ -93,8 +92,7 @@ function Print-Usage() {
   Write-Host "  -help                     Print help and exit"
   Write-Host ""
   Write-Host "Test actions"
-  Write-Host "  -test32                   Run unit tests in the 32-bit runner"
-  Write-Host "  -test64                   Run units tests in the 64-bit runner"
+  Write-Host "  -testArch                 Maps to --arch parameter of dotnet test"
```
Comment (jcouv) [inline eng/build.ps1:95]: Consider documenting the accepted values

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb [hunk 5]
```diff
@@ -13730,7 +13730,7 @@ End Class
 5180801")
         End Sub
 
-        <ConditionalFact(GetType(NoIOperationValidation), GetType(WindowsOnly))>
+        <ConditionalFact(GetType(NoIOperationValidation), GetType(WindowsOnly), GetType(IsEnglishLocal))>
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb:13733]: Why is this needed?

File: src/EditorFeatures/CSharpTest2/EmbeddedLanguages/Json/CSharpJsonTests.cs (hunk 3 above)
Comment (jaredpar) [reply src/EditorFeatures/CSharpTest2/EmbeddedLanguages/Json/CSharpJsonTests.cs:391]: This is a tolerance test and making sure we error vs. crash when we hit the underlying stack limits. Previously the tests were only running on x86 and now they actually run on x64. The tolerance there is higher hence the test needed to be updated.

File: eng/build.ps1 (hunk 4 above)
Comment (jaredpar) [reply eng/build.ps1:95]: I wanted to leave that out because the set of values isn't constant. It's whatever the underlying `dotnet test` supports and that changes across versions.

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb:13733]: Now that we are actually running on x64 this stress test is failing on Spanish x64. The tolerance on Spanish x64 is stronger than English x64 for reasons that I don't fully understand. 

At the moment I'm just trying to get this PR green. This test doesn't change code, it just makes our x64 tests actually run on x64. Fixing the easy bugs I find but overall just trying to get us to green. If this gets us to green my intent is going to file a bug here and merge.

**Comment by jcouv** [comment]
> broken our unit tests

How broken were they?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5)

**Comment by jaredpar** [comment]
@jcouv 

> How broken were they?

It more broke our infra that expected to be running tests on a variety of x86 and x64. It did allow a handful of tests to be written that executed correctly on x86 but failed on x64. Thus far all test bugs, no product bugs. Nevertheless it's tedious to work through the failures to get to a passing state again.
